### PR TITLE
Fixes #53290, puddles are now infinite again.

### LINKED ISSUE #53290: Puddles(such as the one the ash walkers use) are no longer infinite and don't regenerate water
Issue reported from  Round ID: [145233](https://scrubby.melonmesa.com/round/145233) (/tg/Station Basil [ENGLISH] [US-WEST] [100% FREE LAG])

Reporting client version: 513.1524

Puddles are no longer infinite and don't regenerate water. They currently hold a large beaker's worth of water and thats all you get. This is not enough water if you plan on growing multiple plants to maturity as a hermit/ashwalker.

Steps to reproduce:

1. Use 100u of water from a puddle

2. Realise it is now a limited source of water as of #52865

3. Watch as all your plants dry and die cause your an ashwalker and you have no other way to get water

### PR DESCRIPTION
**Pull request by Jack7D1**
## About The Pull Request
Fixes #53290 
A code oversight of #52865 that also applied plumbing changes to puddles.
Avoided snow flaking by creating a new object type called water_source. Can be used in the future instead of the sink type.
Babies first PR

## Why It's Good For The Game
Ashwalkers no longer have to wait for puddles to refill like station sinks.

## Changelog
:cl:
fix: Puddles are now infinite sources of water again.
/:cl:

### CODE REVIEW

File: code/game/objects/structures/watercloset.dm [hunk 1]
```diff
@@ -480,6 +459,175 @@
 		return
 	return ..()
 
+//Legacy sink and puddles, use the type legacysink for unlimited water sources like classic sinks.
```
Comment (Qustinnus) [inline code/game/objects/structures/watercloset.dm:462]: Seeing as the application in this case isn't actually sinks anymore, but the idea of an endless source of a certain reagent, maybe it'd be better to call this type water_source, so it doesn't seem as deprecated

File: code/game/objects/structures/watercloset.dm [hunk 2]
```diff
@@ -480,6 +459,175 @@
...
+	if(!iscarbon(user))
+		return
+	if(!Adjacent(user))
+		return
+
+	if(busy)
+		to_chat(user, "<span class='warning'>Someone's already washing here!</span>")
+		return
+	var/selected_area = parse_zone(user.zone_selected)
+	var/washing_face = 0
```
Comment (Qustinnus) [inline code/game/objects/structures/watercloset.dm:489]:
```suggestion
	var/washing_face = FALSE
```

File: code/game/objects/structures/watercloset.dm [hunk 3]
```diff
@@ -480,6 +459,175 @@
...
+	if(selected_area in list(BODY_ZONE_HEAD, BODY_ZONE_PRECISE_MOUTH, BODY_ZONE_PRECISE_EYES))
+		washing_face = 1
```
Comment (Qustinnus) [inline code/game/objects/structures/watercloset.dm:491]:
```suggestion
		washing_face = TRUE
```

File: code/game/objects/structures/watercloset.dm [hunk 4]
```diff
@@ -480,6 +459,175 @@
...
+		return
+	if(O.item_flags & ABSTRACT) //Abstract items like grabs won't wash. No-drop items will though because it's still technically an item in your hand.
+		return
+
+	if(user.a_intent != INTENT_HARM)
+		to_chat(user, "<span class='notice'>You start washing [O]...</span>")
+		busy = TRUE
+		if(!do_after(user, 40, target = src))
+			busy = FALSE
+			return 1
```
Comment (Qustinnus) [inline code/game/objects/structures/watercloset.dm:577]:
```suggestion
			return TRUE
```

File: code/game/objects/structures/watercloset.dm [hunk 5]
```diff
@@ -480,6 +459,175 @@
...
+		busy = FALSE
+		O.wash(CLEAN_WASH)
+		O.acid_level = 0
+		create_reagents(5)
+		reagents.add_reagent(dispensedreagent, 5)
+		reagents.expose(O, TOUCH)
+		user.visible_message("<span class='notice'>[user] washes [O] using [src].</span>", \
+							"<span class='notice'>You wash [O] using [src].</span>")
+		return 1
```
Comment (Qustinnus) [inline code/game/objects/structures/watercloset.dm:586]:
```suggestion
		return TRUE
```

File: code/game/objects/structures/watercloset.dm (hunk 1 above)
Comment (Qustinnus) [reply code/game/objects/structures/watercloset.dm:462]: also please use autodoc, you can find info on it here https://github.com/SpaceManiac/SpacemanDMM/blob/master/src/dmdoc/README.md

Comment (Jack7D1) [reply code/game/objects/structures/watercloset.dm:462]: A good idea! making the change

File: code/game/objects/structures/watercloset.dm [hunk 6]
```diff
@@ -480,6 +459,175 @@
...
+
+/obj/structure/legacysink/puddle/attackby(obj/item/O, mob/user, params)
+	icon_state = "puddle-splash"
+	. = ..()
+	icon_state = "puddle"
+
+/obj/structure/legacysink/puddle/deconstruct(disassembled = TRUE)
+	qdel(src)
+
+/obj/structure/legacysink/greyscale
```
Comment (Qustinnus) [inline code/game/objects/structures/watercloset.dm:623]: this is only useful for crafting/creating material datum based sinks, seeing as the use-case for these isn't crafted sinks you dont need a greyscale type

File: code/game/objects/structures/watercloset.dm [hunk 7]
```diff
@@ -480,6 +459,175 @@
...
+			playsound(src, B.stun_sound, 50, TRUE)
+			return
+
+	if(istype(O, /obj/item/mop))
+		O.reagents.add_reagent(dispensedreagent, 5)
+		to_chat(user, "<span class='notice'>You wet [O] in [src].</span>")
+		playsound(loc, 'sound/effects/slosh.ogg', 25, TRUE)
+		return
+
+	if(O.tool_behaviour == TOOL_WRENCH && !(flags_1&NODECONSTRUCT_1))
```
Comment (Qustinnus) [inline code/game/objects/structures/watercloset.dm:549]: this allows you to deconstruct this type which doesn't sound like something you'd want

**Comment by Jack7D1** [comment]
Thanks a ton for your help Qustinnus!

File: code/game/objects/structures/watercloset.dm [hunk 8]
```diff
@@ -480,6 +459,151 @@
...
+	var/dispensedreagent = /datum/reagent/water // for whenever plumbing happens
+
+/obj/structure/water_source/attack_hand(mob/living/user)
+	. = ..()
+	if(.)
+		return
+	if(!user || !istype(user))
+		return
+	if(!iscarbon(user))
+		return
```
Comment (Rohesie) [inline code/game/objects/structures/watercloset.dm:479]:
```suggestion
	if(!iscarbon(user))
		return
```
`iscarbon` will filter the above two conditions already

File: code/game/objects/structures/watercloset.dm [hunk 9]
```diff
@@ -480,6 +459,151 @@
...
+	var/selected_area = parse_zone(user.zone_selected)
+	var/washing_face = FALSE
+	if(selected_area in list(BODY_ZONE_HEAD, BODY_ZONE_PRECISE_MOUTH, BODY_ZONE_PRECISE_EYES))
+		washing_face = TRUE
+	user.visible_message("<span class='notice'>[user] starts washing [user.p_their()] [washing_face ? "face" : "hands"]...</span>", \
+						"<span class='notice'>You start washing your [washing_face ? "face" : "hands"]...</span>")
+	busy = TRUE
+
+	if(!do_after(user, 40, target = src))
```
Comment (Rohesie) [inline code/game/objects/structures/watercloset.dm:494]:
```suggestion
	if(!do_after(user, 4 SECONDS, target = src))
```
Favor using the time defines.

File: code/game/objects/structures/watercloset.dm [hunk 10]
```diff
@@ -480,6 +459,151 @@
...
+
+	if(istype(O, /obj/item/reagent_containers))
+		var/obj/item/reagent_containers/RG = O
+		if(RG.is_refillable())
+			if(!RG.reagents.holder_full())
+				RG.reagents.add_reagent(dispensedreagent, min(RG.volume - RG.reagents.total_volume, RG.amount_per_transfer_from_this))
+				to_chat(user, "<span class='notice'>You fill [RG] from [src].</span>")
+				return TRUE
+			to_chat(user, "<span class='notice'>\The [RG] is full.</span>")
+			return FALSE
```
Comment (Rohesie) [inline code/game/objects/structures/watercloset.dm:527]:
```suggestion
		var/obj/item/reagent_containers/container = O
		if(container.is_refillable())
			if(!container.reagents.holder_full())
				container.reagents.add_reagent(dispensedreagent, min(container.volume - container.reagents.total_volume, container.amount_per_transfer_from_this))
				to_chat(user, "<span class='notice'>You fill [container] from [src].</span>")
				return TRUE
			to_chat(user, "<span class='notice'>\The [container] is full.</span>")
			return FALSE
```
Avoid the usage of a few upper-case letters for variable names. Use clear and informative names instead.

File: code/game/objects/structures/watercloset.dm [hunk 11]
```diff
@@ -480,6 +459,151 @@
...
+	if(istype(O, /obj/item/melee/baton))
+		var/obj/item/melee/baton/B = O
+		if(B.cell && B.cell.charge && B.turned_on)
+			flick("baton_active", src)
+			user.Paralyze(B.stun_time)
+			user.stuttering = B.stun_time/20
+			B.deductcharge(B.cell_hit_cost)
+			user.visible_message("<span class='warning'>[user] shocks [user.p_them()]self while attempting to wash the active [B.name]!</span>", \
+								"<span class='userdanger'>You unwisely attempt to wash [B] while it's still on.</span>")
+			playsound(src, B.stun_sound, 50, TRUE)
```
Comment (Rohesie) [inline code/game/objects/structures/watercloset.dm:538]:
```suggestion
		var/obj/item/melee/baton/baton = O
		if(baton.cell && baton.cell.charge && baton.turned_on)
			flick("baton_active", src)
			user.Paralyze(baton.stun_time)
			user.stuttering = baton.stun_time * 0.05
			baton.deductcharge(baton.cell_hit_cost)
			user.visible_message("<span class='warning'>[user] shocks [user.p_them()]self while attempting to wash the active [baton.name]!</span>", \
								"<span class='userdanger'>You unwisely attempt to wash [baton] while it's still on.</span>")
			playsound(src, baton.stun_sound, 50, TRUE)
```
Favor multiplications over divisions whenever it doesn't impact readability.

File: code/game/objects/structures/watercloset.dm [hunk 12]
```diff
@@ -480,6 +459,151 @@
...
+
+	if(istype(O, /obj/item/mop))
+		O.reagents.add_reagent(dispensedreagent, 5)
+		to_chat(user, "<span class='notice'>You wet [O] in [src].</span>")
+		playsound(loc, 'sound/effects/slosh.ogg', 25, TRUE)
+		return
+
+	if(istype(O, /obj/item/stack/medical/gauze))
+		var/obj/item/stack/medical/gauze/G = O
+		new /obj/item/reagent_containers/glass/rag(src.loc)
```
Comment (Rohesie) [inline code/game/objects/structures/watercloset.dm:549]:
```suggestion
		new /obj/item/reagent_containers/glass/rag(loc)
```
`src` is implied, avoid using it as a prefix unless there's another local variable with the same name.

File: code/game/objects/structures/watercloset.dm [hunk 13]
```diff
@@ -480,6 +459,151 @@
...
+		return
+
+	if(istype(O, /obj/item/stack/ore/glass))
+		new /obj/item/stack/sheet/sandblock(loc)
+		to_chat(user, "<span class='notice'>You wet the sand and form it into a block.</span>")
+		O.use(1)
+		return
+
+	if(!istype(O))
+		return
```
Comment (Rohesie) [inline code/game/objects/structures/watercloset.dm:561]:
```suggestion
```
This condition should never happen.
If there was any chance for this to happen, we should be using `CRASH()` with an informative message, and not returning silently.

File: code/game/objects/structures/watercloset.dm [hunk 14]
```diff
@@ -480,6 +459,151 @@
...
+	if(O.item_flags & ABSTRACT) //Abstract items like grabs won't wash. No-drop items will though because it's still technically an item in your hand.
+		return
+
+	if(user.a_intent != INTENT_HARM)
+		to_chat(user, "<span class='notice'>You start washing [O]...</span>")
+		busy = TRUE
+		if(!do_after(user, 40, target = src))
```
Comment (Rohesie) [inline code/game/objects/structures/watercloset.dm:568]:
```suggestion
		if(!do_after(user, 4 SECONDS, target = src))
```

File: code/game/objects/structures/watercloset.dm [hunk 15]
```diff
@@ -480,6 +459,151 @@
...
+		O.wash(CLEAN_WASH)
+		O.acid_level = 0
+		create_reagents(5)
+		reagents.add_reagent(dispensedreagent, 5)
+		reagents.expose(O, TOUCH)
+		user.visible_message("<span class='notice'>[user] washes [O] using [src].</span>", \
+							"<span class='notice'>You wash [O] using [src].</span>")
+		return TRUE
+	else
+		return ..()
```
Comment (Rohesie) [inline code/game/objects/structures/watercloset.dm:581]:
```suggestion

	return ..()
```
No need for an else her, as there's a `return` at the end of the `if`

File: code/game/objects/structures/watercloset.dm (hunk 8 above)
Comment (Jack7D1) [reply code/game/objects/structures/watercloset.dm:479]: Huh, didn't know that!

File: code/game/objects/structures/watercloset.dm (hunk 9 above)
Comment (Jack7D1) [reply code/game/objects/structures/watercloset.dm:492]: Thanks!

File: code/game/objects/structures/watercloset.dm (hunk 13 above)
Comment (Jack7D1) [reply code/game/objects/structures/watercloset.dm:559]: I dont even remember that being in there

**Pull request merged**